File: sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/LLM/RunAnywhere+LoRA.swift
// RunAnywhere+LoRA.swift
// RunAnywhere SDK
//
// Public API for LoRA adapter management — namespaced under
// `RunAnywhere.lora.*` per the canonical cross-SDK spec
// (CANONICAL_API §3 — LoRA).
//
// Runtime operations delegate to CppBridge.LLM; catalog operations
// delegate to CppBridge.LoraRegistry.

import Foundation

// MARK: - LoRA Capability Namespace

public extension RunAnywhere {

    /// Capability accessor for LoRA adapter management.
    ///
    /// Mirrors the namespaced `lora.*` shape used by the other SDKs
    /// (Kotlin/Flutter/RN/Web). All eight canonical methods live on
    /// the returned `LoRA` value.
    static var lora: LoRA { LoRA() }

    /// Stateless namespace exposing the canonical 8-method LoRA surface.
    /// Backed by the C ABI via `CppBridge.LLM` (runtime ops) and
    /// `CppBridge.LoraRegistry` (catalog ops).
    struct LoRA: Sendable {

        fileprivate init() {}

        // MARK: Runtime Operations

        /// Load and apply a LoRA adapter to the currently loaded model.
        /// Multiple adapters can be stacked. Context is recreated internally.
        ///
        /// - Parameter config: `LoRAAdapterConfig` with path and scale.
        /// - Returns: `LoRAAdapterInfo` describing the loaded adapter.
        @discardableResult
        public func load(_ config: LoRAAdapterConfig) async throws -> LoRAAdapterInfo {
            let info = try await load(config.toRALoRAAdapterConfig())
            return LoRAAdapterInfo(from: info)
        }

        /// Load and apply a generated-proto LoRA adapter config.
        @discardableResult
        public func load(_ config: RALoRAAdapterConfig) async throws -> RALoRAAdapterInfo {
            guard RunAnywhere.isInitialized else {
                throw SDKException.general(.notInitialized, "SDK not initialized")
            }
            return try await CppBridge.LLM.shared.loadLoraAdapter(config)
        }

        /// Remove a specific LoRA adapter by adapter id (file path).
        public func remove(_ adapterId: String) async throws {
            guard RunAnywhere.isInitialized else {
                throw SDKException.general(.notInitialized, "SDK not initialized")
            }
            var config = RALoRAAdapterConfig()
            config.adapterPath = adapterId
            config.adapterID = adapterId
            _ = try await CppBridge.LLM.shared.removeLoraAdapter(config)
        }

        /// Remove all loaded LoRA adapters.
        public func clear() async throws {
            guard RunAnywhere.isInitialized else {
                throw SDKException.general(.notInitialized, "SDK not initialized")
            }
            _ = try await CppBridge.LLM.shared.clearLoraAdaptersProto()
        }

        /// Get info about all currently loaded LoRA adapters.
        public func getLoaded() async throws -> [LoRAAdapterInfo] {
            guard RunAnywhere.isInitialized else {
                throw SDKException.general(.notInitialized, "SDK not initialized")
            }
            return try await CppBridge.LLM.shared.getLoadedLoraAdapters()
        }

        /// Check whether a LoRA adapter is compatible with a model.
        ///
        /// `adapterId` is the path to the adapter GGUF (matching the C ABI
        /// `rac_llm_check_lora_compatibility`). `modelId` is reserved for
        /// future use by the canonical API; the current C ABI checks
        /// against the active model loaded via `loadLLMModel`.
        public func checkCompatibility(
            adapterId: String,
            modelId _: String
        ) async -> LoraCompatibilityResult {
            guard RunAnywhere.isInitialized else {
                return LoraCompatibilityResult(isCompatible: false, error: "SDK not initialized")
            }
            var config = RALoRAAdapterConfig()
            config.adapterPath = adapterId
            do {
                let proto = try await CppBridge.LLM.shared.checkLoraCompatibility(config)
                return LoraCompatibilityResult(from: proto)
            } catch {
                return LoraCompatibilityResult(isCompatible: false, error: error.localizedDescription)
            }
        }

        // MARK: Catalog Operations

        /// Register a LoRA adapter config in the SDK catalog at app startup (CANONICAL_API §3).
        ///
        /// Call this before loading any adapters so the SDK knows what's available.
        /// The `config` carries the adapter path and scale; catalog metadata
        /// (name, description, etc.) is stored internally.
        ///
        /// - Parameter config: `LoRAAdapterConfig` with path and scale.
        public func register(_ config: LoRAAdapterConfig) async throws {
            // Wrap the config into a catalog entry using the path as the id/name.
            let entry = LoraAdapterCatalogEntry(
                id: config.path,
                name: config.path,
                description: "Registered via lora.register(config:)",
                downloadURL: URL(fileURLWithPath: config.path),
                filename: (config.path as NSString).lastPathComponent,
                compatibleModelIds: [],
                fileSize: 0,
                defaultScale: config.scale
            )
            try await register(entry)
        }

        /// Register a LoRA adapter from a full catalog entry.
        ///
        /// This overload accepts `LoraAdapterCatalogEntry` directly, which carries
        /// richer metadata (name, description, download URL, compatible model IDs).
        ///
        /// - Parameter entry: A complete `LoraAdapterCatalogEntry` to register.
        public func register(_ entry: LoraAdapterCatalogEntry) async throws {
            try await register(entry.toRALoraAdapterCatalogEntry())
        }

        /// Register a generated-proto LoRA catalog entry.
        public func register(_ entry: RALoraAdapterCatalogEntry) async throws {
            guard RunAnywhere.isInitialized else {
                throw SDKException.general(.notInitialized, "SDK not initialized")
            }
            _ = try await CppBridge.LoraRegistry.shared.register(entry)
        }

        /// Get all LoRA adapters compatible with a specific model (CANONICAL_API §3).
        ///
        /// - Parameter modelId: Model identifier to filter by.
        /// - Returns: `[LoRAAdapterInfo]` for compatible adapters.
        public func adaptersForModel(_ modelId: String) async -> [LoRAAdapterInfo] {
            let entries = await CppBridge.LoraRegistry.shared.getForModel(modelId)
            return entries.map { entry in
                LoRAAdapterInfo(path: entry.id, scale: entry.defaultScale, applied: false)
            }
        }

        /// Get all registered LoRA adapters (CANONICAL_API §3).
        ///
        /// - Returns: `[LoRAAdapterInfo]` for all registered adapters.
        public func allRegistered() async -> [LoRAAdapterInfo] {
            let entries = await CppBridge.LoraRegistry.shared.getAll()
            return entries.map { entry in
                LoRAAdapterInfo(path: entry.id, scale: entry.defaultScale, applied: false)
            }
        }
    }
}

File: sdk/runanywhere-react-native/packages/core/src/Public/Extensions/RunAnywhere+LoRA.ts
/**
 * RunAnywhere+LoRA.ts
 *
 * Public API for LoRA adapter management. Wave 2: aligned to
 * proto-canonical LoRA shapes (`@runanywhere/proto-ts/lora_options`).
 *
 * Matches Swift: `Public/Extensions/LLM/RunAnywhere+LoRA.swift`. Surface
 * follows the canonical cross-SDK namespace shape — `RunAnywhere.lora.*`:
 *
 *   await RunAnywhere.lora.load({ adapterPath, scale })
 *   await RunAnywhere.lora.remove(adapterPath)
 *   await RunAnywhere.lora.clear()
 *   const loaded = await RunAnywhere.lora.getLoaded()
 *   const compat = await RunAnywhere.lora.checkCompatibility(adapterPath)
 *   await RunAnywhere.lora.register({ id, name, ... })
 *   const list = await RunAnywhere.lora.adaptersForModel(modelId)
 *   const all  = await RunAnywhere.lora.allRegistered()
 */

import { requireNativeModule, isNativeModuleAvailable } from '../../native';
import { SDKLogger } from '../../Foundation/Logging/Logger/SDKLogger';
import type {
  LoRAAdapterConfig,
  LoRAAdapterInfo,
  LoraCompatibilityResult,
} from '@runanywhere/proto-ts/lora_options';
import {
  LoRAAdapterConfig as LoRAAdapterConfigMessage,
  LoRAAdapterInfo as LoRAAdapterInfoMessage,
  LoraCompatibilityResult as LoraCompatibilityResultMessage,
} from '@runanywhere/proto-ts/lora_options';
import {
  arrayBufferToBytes,
  bytesToArrayBuffer,
} from '../../services/ProtoBytes';

const logger = new SDKLogger('RunAnywhere.LoRA');

function encodeConfig(config: LoRAAdapterConfig): ArrayBuffer {
  return bytesToArrayBuffer(LoRAAdapterConfigMessage.encode(
    LoRAAdapterConfigMessage.create(config)
  ).finish());
}

function decodeAdapterInfo(buffer: ArrayBuffer, operation: string): LoRAAdapterInfo {
  const bytes = arrayBufferToBytes(buffer);
  if (bytes.byteLength === 0) {
    throw new Error(`${operation} returned an empty LoRA proto result`);
  }
  return LoRAAdapterInfoMessage.decode(bytes);
}

// ============================================================================
// Runtime Operations
// ============================================================================

/**
 * Load and apply a LoRA adapter to the currently loaded model.
 *
 * Returns `LoRAAdapterInfo` describing the applied adapter, as specified by
 * the canonical cross-SDK spec §3.
 *
 * Canonical: `RunAnywhere.lora.load(config) → LoRAAdapterInfo`.
 */
async function load(config: LoRAAdapterConfig): Promise<LoRAAdapterInfo> {
  if (!isNativeModuleAvailable()) {
    throw new Error('Native module not available');
  }
  const native = requireNativeModule();
  const info = decodeAdapterInfo(
    await native.loraLoadProto(encodeConfig(config)),
    'loraLoadProto'
  );
  logger.info(`LoRA adapter loaded: ${config.adapterPath}`);
  return info;
}

/**
 * Remove a specific LoRA adapter by path.
 *
 * Canonical: `RunAnywhere.lora.remove(adapterId)`.
 */
async function remove(adapterId: string): Promise<void> {
  if (!isNativeModuleAvailable()) {
    throw new Error('Native module not available');
  }
  const native = requireNativeModule();
  await native.loraRemoveProto(encodeConfig({
    adapterId,
    adapterPath: adapterId,
    scale: 1.0,
  }));
  logger.info(`LoRA adapter removed: ${adapterId}`);
}

/** Remove all loaded LoRA adapters. Canonical: `RunAnywhere.lora.clear()`. */
async function clear(): Promise<void> {
  if (!isNativeModuleAvailable()) {
    throw new Error('Native module not available');
  }
  await requireNativeModule().loraClearProto();
  logger.info('All LoRA adapters cleared');
}

/**
 * Get info about all currently loaded LoRA adapters.
 *
 * Canonical: `RunAnywhere.lora.getLoaded()`.
 */
async function getLoaded(): Promise<LoRAAdapterInfo[]> {
  return [];
}

/**
 * Check LoRA adapter compatibility with a model.
 *
 * Both `adapterId` and `modelId` are forwarded to the native bridge. When
 * `modelId` is omitted the bridge checks against the currently loaded model.
 *
 * Canonical: `RunAnywhere.lora.checkCompatibility(adapterId, modelId)` (§3).
 */
async function checkCompatibility(
  adapterId: string,
  modelId?: string,
): Promise<LoraCompatibilityResult> {
  if (!isNativeModuleAvailable()) {
    return { isCompatible: false, errorMessage: 'SDK not initialized' };
  }
  const native = requireNativeModule();
  const config = LoRAAdapterConfigMessage.create({
    adapterId,
    adapterPath: adapterId,
    scale: 1.0,
  });
  void modelId;
  const bytes = arrayBufferToBytes(
    await native.loraCompatibilityProto(
      bytesToArrayBuffer(LoRAAdapterConfigMessage.encode(config).finish())
    )
  );
  return bytes.byteLength > 0
    ? LoraCompatibilityResultMessage.decode(bytes)
    : { isCompatible: false, errorMessage: 'LoRA proto ABI unavailable' };
}

// ============================================================================
// Catalog Operations
// ============================================================================

/**
 * Register a LoRA adapter configuration with the SDK so it can be
 * referenced by `adapterId` in subsequent `load` calls.
 *
 * Accepts `LoRAAdapterConfig` per the canonical cross-SDK spec §3.
 *
 * Canonical: `RunAnywhere.lora.register(config: LoRAAdapterConfig) → void`.
 */
async function register(config: LoRAAdapterConfig): Promise<void> {
  void config;
  throw new Error('LoRA catalog registration is not exposed by the RN core bridge because commons does not provide a lifecycle-owned LoRA registry handle.');
}

/**
 * Get all LoRA adapters compatible with a specific model.
 *
 * Canonical: `RunAnywhere.lora.adaptersForModel(modelId) → LoRAAdapterInfo[]` (§3).
 */
async function adaptersForModel(
  modelId: string,
): Promise<LoRAAdapterInfo[]> {
  void modelId;
  return [];
}

/**
 * Get all registered LoRA adapters.
 *
 * Canonical: `RunAnywhere.lora.allRegistered() → LoRAAdapterInfo[]` (§3).
 */
async function allRegistered(): Promise<LoRAAdapterInfo[]> {
  return [];
}

// ============================================================================
// Canonical namespace export
// ============================================================================

/**
 * `RunAnywhere.lora` namespace — canonical cross-SDK shape.
 *
 * Mirror of Swift `RunAnywhere.lora.load(...)` and Web/Flutter/Kotlin
 * once the same alignment lands there. Stateless wrapper: each call
 * dispatches into the LLM component native bridge.
 */
export const lora = {
  load,
  remove,
  clear,
  getLoaded,
  checkCompatibility,
  register,
  adaptersForModel,
  allRegistered,
};
